### ISSUE TITLE
wav2vec-u how to get close to the LibriSpeech WERs after GAN training

### ISSUE THREAD

**Issue by JeromeNi** [facebookresearch/fairseq#3617]
### Before asking:   
1. search the issues.   
2. search the docs.    

#### What is your question?
Hello, I'm having trouble getting close to the reported WER for LibriSpeech after GAN training but before self-training (i.e., third-to-last row of Table 2; https://ai.facebook.com/research/publications/unsupervised-speech-recognition/). As processing 960h of LibriSpeech takes too much storage and time, and that from Figure 8 the AM resulted from GAN training is not largely affected by audio corpus size, I only used the train-clean-100 subset for GAN training. The phone and word language models, on the other hand, takes the full LibriSpeech LM corpus (openslr 11). Phonemization was done using g2p_en.

Without the pykaldi decoder for integrating a phone WFST produced by `prepare_text.sh` I could get around 27% PER on dev-other (using features from xxx_mean_pooled); with the phone LM, it would decrease down to around 23% (using features from xxx_mean without pooled). With the pykaldi decoder for integrating a word WFST (using features from xxx_mean without pooled), the WER however, is around 38%.
 
So my questions are:
1. Is Figure 8 of the paper produced with a phone LM? If so, it would seem that the PER I got was only slightly higher. If not is there any way to further reduce the PER?
2. Are the WERs in the third-to-last row of Table 2 produced with phoneme emission probabilities from the GAN model, or do those come from the self-trained phone HMMs?
3. From Section 5.4 (Decoding) of the paper, it seems that decoding involves only 2 hyperparameters, `acoustic_scale` in `kaldi_decoder_config` as well as `blank_weight`. However, if I set blank weight anywhere other than zero (such as -1,1), I immediately get much higher WER. I have attached my decoding configs below.
4. I found that for some reason, in line 585 of the script `w2vu_generate.py` the flag `"no_softmax": True` is producing a much higher WER. Also, should the condition on line 580 be `unpaired_audio_text`, the task name of GAN training, instead?
5. What else would you suggest to get close to the reported WER after GAN training?

Thanks!!!

#### Code

  

Below is the yaml I created for the pykaldi phone decoder:

> w2l_decoder: KALDI
post_process: silence
blank_weight: 0
sil_is_blank: true
blank_mode: add
unsupervised_tuning: 0 # tuning mode is 0
viterbi_transcript: /path/to/wav2vec_pseudo_transcriptions/dev_other.txt
min_vt_uer: 0.03
min_lm_ppl: 7.842956636796915
targets: phn
kaldi_decoder_config:
    acoustic_scale: 4
    hlg_graph_path: /path/to/unpaired_text_libri/fst/phn_to_phn_sil/HLG.phn.lm.phones.filtered.06.fst
    output_dict: /path/to/unpaired_text_libri/fst/phn_to_phn_sil/kaldi_dict.lm.phones.filtered.06.txt
results_path: /path/to/wav2vec_pseudo_transcriptions_kaldi_phn/

Below is the yaml I have for pykaldi word decoder:

> w2l_decoder: KALDI
post_process: silence
blank_weight: 0
sil_is_blank: true
blank_mode: add
unsupervised_tuning: 0 # tuning mode is 0
viterbi_transcript: /path/to/wav2vec_pseudo_transcriptions/dev_other.txt
lexicon: /path/to/unpaired_text_libri/lexicon_filtered.lst
min_vt_uer: 0.03
min_lm_ppl: 7.842956636796915 # by running the command query on the phone LM
targets: wrd
kaldi_decoder_config:
    acoustic_scale: 3
    hlg_graph_path: /path/to/unpaired_text_libri/fst/phn_to_words_sil/HLG.phn.kenlm.wrd.o40003.fst
    output_dict: /path/to/unpaired_text_libri/fst/phn_to_words_sil/kaldi_dict.kenlm.wrd.o40003.txt
    results_path: /path/to/wav2vec_pseudo_transcriptions_kaldi_wrd/

#### What have you tried?
Tuned the acoustic scale and blank weight, etc. Details are above.

#### What's your environment?

 - fairseq Version (e.g., 1.0 or master): master
 - PyTorch Version (e.g., 1.0): 1.7.1
 - OS (e.g., Linux): Linux
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source): 
 - Python version: 3.7.9
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: V100
 - Any other relevant information: N/A

**Comment by alexeib** [comment]
1. it is without phone lm. i just recently reran the pipeline using 10h of audio only and using 100h transcriptions as text data and got something like 18% PER before any LM. are you sure you followed the preprocessing steps correctly and you are using the g2p phonemizer for librispeech text data (espeak phonemizer will have worse results, similar to what you are getting)
2. those are with a gan model with 4gram word LM (used via kaldi fsts). dont forget that when decoding with an LM you want to use features from the step before mean pooling adjacent timesteps
3. we tune with a hyperparam tuner such as Ax or Nevergrad, which sometimes chooses blank weight close to 0, and sometimes not. this depends on the scale of the weights your model has learned. there is nothing wrong with using a blank weight of 0, you just might not get the most optimal results (but probably the diff wont be very big)
4. good catch, line 580 references incorrect task - you are right. softmax is unnecessary for decoding (decoder will normalize emissions itself if neessary) - the important thing that block does is disable in-model segmentation. will update the code soon.

if you have .phn labels for your dev set you will dev valid_uer during the gan training. it is usually a pretty reliable proxy to model quality. what are you getting there?

**Comment by JeromeNi** [comment]
>     1. it is without phone lm. i just recently reran the pipeline using 10h of audio only and using 100h transcriptions as text data and got something like 18% PER before any LM. are you sure you followed the preprocessing steps correctly and you are using the g2p phonemizer for librispeech text data (espeak phonemizer will have worse results, similar to what you are getting)
> 
>     2. those are with a gan model with 4gram word LM (used via kaldi fsts). dont forget that when decoding with an LM you want to use features from the step before mean pooling adjacent timesteps

1. Yes, I'm using the `g2p_en`, not `espeak`.
2. Yes, I'm using the features in the folder with the suffix `_mean`. I only used the features for `mean_pooled` during training and viterbi decoding.

> if you have .phn labels for your dev set you will dev valid_uer during the gan training. it is usually a pretty reliable proxy to model quality. what are you getting there?
I'm getting around the same uer as viterbi decoding without LM, i.e., 26-28% PER.

The only two settings I have changed (from the provided yaml) is to increase  distributed_world_size to 4 and decrease batch_size to 40 so that the effective batch size of training on 4 GPUs is still 160. I have also changed the following `mean_pool_join:` from `false` to `true`. From what I understood from the code it takes the mean over segments instead of random sampling.

**Comment by alexeib** [comment]
yes, but random sampling works better during training. have you tried leaving mean_pool_join at false?

**Comment by mahshidaln** [comment]
> 1. it is without phone lm. i just recently reran the pipeline using 10h of audio only and using 100h transcriptions as text data and got something like 18% PER before any LM. are you sure you followed the preprocessing steps correctly and you are using the g2p phonemizer for librispeech text data (espeak phonemizer will have worse results, similar to what you are getting)

Hi @alexeib, 
Did you make any significant changes to the hyperparameters when training the model on such amount of data?

**Comment by alexeib** [comment]
no i was using the exact same params

**Comment by mahshidaln** [comment]
> no i was using the exact same params

Thanks! Any idea what I might have missed that the gan training is not converging? (I followed the data preprocessing steps as mentioned on the readme)

**Comment by xiabingquan** [comment]
Hi, @alexeib . In Figure 8, you show that the model can still converge well when trained on a much smaller amount of data. I'm wondering whether you've done similar ablation studies in Wav2Vec-U 2.0. I didn't see any similar ablation study on the scale of dataset in the paper \<\<Towards End-to-end Unsupervised Speech Recognition\>\>. Thanks!!!

**Comment by alexeib** [comment]
we haven't done such ablations but since wav2vec-u 2.0 builds on top of wav2vec-u and is mostly about simplifying pre-processing while improving accuracy, i would be very surprised if a similar ablation with w2v-u 2.0 did not follow the same trends as in 1.0
